lib/fluent/plugin/out_http.rb

```diff
@@ -244,15 +260,45 @@ def create_request(chunk, uri)
       req
     end
 
+    def make_request_cached(uri, req)
+      id = Thread.current.thread_variable_get(plugin_id)
+      if id.nil?
+        @cache_id_mutex.synchronize {
+          id = @cache_id
+          @cache_id += 1
+        }
+        Thread.current.thread_variable_set(plugin_id, id)
+      end
+      uri_str = uri.to_s
+      if @cache[id].uri != uri_str
+        @cache[id].conn.finish if @cache[id].conn&.started?
+        http =  if @proxy_uri
+                  Net::HTTP.start(uri.host, uri.port, @proxy_uri.host, @proxy_uri.port, @proxy_uri.user, @proxy_uri.password, @http_opt)
+                else
+                  Net::HTTP.start(uri.host, uri.port, @http_opt)
+                end
+        @cache[id] = @cache_entry.new(uri_str, http)
+      end
+      @cache[id].conn.request(req)
+    end
```

Comment:
Sorry for the overall change, but I believe we can use more simple logic like this.

```suggestion
    def make_request_cached(uri, req)
      thread_name = Thread.current.name

      @cache_by_thread[thread_name] ||= CacheEntry.new(uri.to_s, make_request(uri, req))
      cache = @cache_by_thread[thread_name]

      unless cache.uri == uri.to_s
        cache.conn.finish if cache.conn.started?
        @cache_by_thread[thread_name] = CacheEntry.new(uri.to_s, make_request(uri, req))
        cache = @cache_by_thread[thread_name]
      end

      cache.conn.request(req)
    end
```

Note: we can use the thread name `flush_thread_{i}` for the key.

https://github.com/fluent/fluentd/blob/0893c3972297c103067499dd55dc538e6a21a4a1/lib/fluent/plugin/output.rb#L506-L511